README.md

```diff
@@ -1,486 +1,679 @@
+`@logged` can be implemented in JavaScript as a decorator. Decorators are functions that are called with an argument containing what's being decorated. For example:
+- A decorated method is called with the method being decorated
+- A decorated getter is called with the getter function being decorated
+- A decorated setter is called with the setter function being decorated
+
+(Decorators are called with a second parameter giving more context, but we don't need those details for the `@logged` decorator.)
+
+The return value of a decorator is a new value that replaces the thing it's wrapping. For methods, getters and setters, the return value is another function to replace that method, getter or setter.
 
 ```mjs
 // logged.mjs
 
-export decorator @logged {
-  @wrap(f => {
-    const name = f.name;
-    function wrapped(...args) {
-      console.log(`starting ${name} with arguments ${args.join(", ")}`);
-      f.call(this, ...args);
-      console.log(`ending ${name}`);
-    }
-    wrapped.name = name;
-    return wrapped;
-  })
+export function logged(f) {
+  const name = f.name;
+  function wrapped(...args) {
+    console.log(`starting ${name} with arguments ${args.join(", ")}`);
+    const ret = f.call(this, ...args);
+    console.log(`ending ${name}`);
+    return ret;
+  }
+  wrapped.name = name;
+  return wrapped;
 }
 ```
 
-In the above example, the *composed decorator* `@logged` is defined to expand out into a call of the `@wrap` decorator with a particular fixed callback.
+This example roughly "desugars" to the following (i.e., could be transpiled as such):
 
-### `@defineElement`
+```js
+let x_setter;
+
+class C {
+  method(arg) {
+    this.#x = arg;
+  }
+
+  static #x_setter(value) { }
+  static #_ = (x_setter = C.#x_setter, void 0);
+  set #x(value) { return x_setter.call(this, value); }
+}
+
+C.prototype.method = logged(C.prototype.method, { kind: "method", name: "method", isStatic: false });
+x_setter = logged(x_setter, {kind: "setter", isStatic: false});
+```
+
+Note that getters and setters are decorated separately. Accessors are not "coalesced" as in earlier decorators proposals  (unless they are generated for a field; see below).
+
+## `@defineElement`
 
 [HTML Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) lets you define your own HTML element. Elements are registered using `customElements.define`. Using decorators, the registration can happen up-front:
 
 ```mjs
-import { @defineElement } from "./defineElement.mjs";
+import { defineElement } from "./defineElement.mjs";
 
 @defineElement('my-class')
 class MyClass extends HTMLElement { }
 ```
 
-The `@defineElement` decorator is based on the `@register` decorator. This decorator is called when the class is finished being defined, and it calls the provided callback with the class that resulted.
+Classes can be decorated just like methods and accessors. The class shows up in the `value` option.
```

Comment:
Doesn't class decorator still compromise static analyzability? What if instead of returning a new class the user decides to modify the existing one?

Reply:
It could, but this proposal is more analyzable than some past proposals where everything happened by modifying the class. The important thing is to have this property in practice.

Reply:
Hmm okay. Well, if it needs to be analyzable all the way to be adopted by TypeScript and the like, the only easy way I can think of to let users extend the class but not modify it is to either freeze the class and its prototype before passing to decorators, so now all decorated classes would be frozen, or wrap it in a proxy before passing as an argument to decorators. Then you could guarantee analyzability.

Reply:
On a slightly different note, is metadata from property and method decorators available to class decorators or does it get assigned after all decorator function have been executed?

Reply:
@dolsem I don't think we should do this freezing by default, as it would make decorated classes incompatible with some use cases. In general, this proposal isn't trying to make everything about classes totally statically analyzable, just *the use of decorators itself*.

Yes, metadata from field and method decorators is available to class decorators.